### PR TITLE
WinUI 3/islands support

### PR DESCRIPTION
**Pull request by asklar**
This relies on bits that will only be available publicly via WinUI 3 preview 3, so for now I'm using a nuget package from CI. 

All of the changes are usable with system xaml and winui 3 preview 2 _except_ for one in `TouchEventHandler` (since the IXP apis are changing), so that one change is behind a `WINUI3_PREVIEW3` preprocessor define.

The facebook api for native UI manager has a fixed set of exception types it handles, and obviously those don't include `winrt::hresult_error`, so when a WinRT type activation fails, a `hresult_error` is thrown and goes unhandled crashing the whole process, so adding a conversion from `hresult_error` to the error message so that the frame that's lower in the stack handles it (since it catches among a few other things, `std::string`s).

I'm also including a script to transform packages.config and winui.props automatically based on a new nuget package.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5576)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/packages.config [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
+<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="boost" version="1.72.0.0" targetFramework="native" />
   <package id="Microsoft.Build.Tasks.Git" version="1.0.0" targetFramework="native" developmentDependency="true" />
   <package id="Microsoft.SourceLink.Common" version="1.0.0" targetFramework="native" developmentDependency="true" />
   <package id="Microsoft.SourceLink.GitHub" version="1.0.0" targetFramework="native" developmentDependency="true" />
   <package id="Microsoft.UI.Xaml" version="2.3.191129002" targetFramework="native" />
   <package id="Microsoft.Windows.CppWinRT" version="2.0.200615.7" targetFramework="native" />
-  <package id="Microsoft.WinUI" version="3.0.0-preview2.200713.0" targetFramework="native" />
+  <package id="Microsoft.WinUI" version="3.0.0-preview2.200713.0" targetFramework="native"/>
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/packages.config:9]: >"/> [](start = 89, length = 3)

This change is unnecessary. Please keep the same formatting as for the other lines.

File: vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp [hunk 2]
```diff
@@ -30,14 +32,17 @@ void DeviceInfoHolder::InitDeviceInfoHolder(
...
     auto const &displayInfo = winrt::Windows::Graphics::Display::DisplayInformation::GetForCurrentView();
-    auto const &window = xaml::Window::Current().CoreWindow();
-
-    deviceInfoHolder->m_sizeChangedRevoker =
-        window.SizeChanged(winrt::auto_revoke, [weakHolder = std::weak_ptr(deviceInfoHolder)](auto &&, auto &&) {
-          if (auto strongHolder = weakHolder.lock()) {
-            strongHolder->updateDeviceInfo();
-          }
-        });
+    /// TODO: WinUI 3 islands support
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp:35]: >/// [](start = 4, length = 3)

Please use double slash for normal comments. The triple slash line comments are usually used for automated docs.

Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp:35]: >WinUI 3 islands support [](start = 14, length = 23)

Please add a verb. It is not clear what type of action needs to be done here.

File: vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp [hunk 3]
```diff
@@ -86,14 +91,28 @@ void DeviceInfoHolder::SetCallback(
...
+    auto const window = xaml::Window::Current().CoreWindow();
+
+    m_windowWidth = window.Bounds().Width;
+    m_windowHeight = window.Bounds().Height;
+  } else {
+    /// TODO: WinUI 3 Island - mock for now
+    m_windowWidth = 600;
+    m_windowHeight = 800;
+
+    /*
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp:104]: >/* [](start = 4, length = 2)

We typically do not add commented out code.

File: vnext/Shared/Modules/UIManagerModule.cpp [hunk 4]
```diff
@@ -248,20 +249,24 @@ void UIManager::createView(
...
+
+    node->createView();
+    m_nativeUIManager->CreateView(*node, props);
+
+    m_nodeRegistry.addNode(shadow_ptr(node), tag);
+
+    if (!props.isNull())
+      node->updateProperties(std::move(props));
+  } catch (winrt::hresult_error &hr) {
+    throw Microsoft::Common::Unicode::Utf16ToUtf8(hr.message());
```
Comment (vmoroz) [inline vnext/Shared/Modules/UIManagerModule.cpp:268]: >throw [](start = 4, length = 5)

It is considered as an anti-pattern in the modern C++ to throw simple types. The recommended approach is to always throw exceptions inherited from the std::exception.
Why is it needed here?

File: packages/playground/windows/Playground-win32 (Package)/Playground-win32 (Package).wapproj [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (vmoroz) [inline packages/playground/windows/Playground-win32 (Package)/Playground-win32 (Package).wapproj:1]: Could we avoid having spaces in the folder and file names? Many tools will fail to work. Especially if they originally created for Unix-like OSes.

File: packages/playground/windows/playground-win32/Playground-win32.vcxproj [hunk 6]
```diff
@@ -10,6 +10,18 @@
     <WindowsTargetPlatformMinVersion>10.0.18362.0</WindowsTargetPlatformMinVersion>
     <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
     <BuildMSRNCxx>false</BuildMSRNCxx>
+    <RunAutolinkCheck>false</RunAutolinkCheck>
+  </PropertyGroup>
+  <PropertyGroup Label="FromWinUI3_VSIX" Condition="'$(UseWinUI3)'=='true'">
```
Comment (vmoroz) [inline packages/playground/windows/playground-win32/Playground-win32.vcxproj:15]: >PropertyGroup [](start = 3, length = 13)

Could we make most of these settings to be in the "Global" group above?

File: packages/playground/windows/playground-win32/Playground-Win32.h [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+int __stdcall WinMain(HINSTANCE instance, HINSTANCE, PSTR, int showCmd);
```
Comment (vmoroz) [inline packages/playground/windows/playground-win32/Playground-Win32.h:3]: >WinMain [](start = 14, length = 7)

Why is it needed?

File: packages/playground/windows/playground-win32/Playground-Win32.h [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (vmoroz) [inline packages/playground/windows/playground-win32/Playground-Win32.h:1]: >#pragma [](start = 0, length = 7)

The file is missing the copyright header.

File: packages/playground/windows/Playground-win32 (Package)/Playground-win32 (Package).wapproj (hunk 5 above)
Comment (asklar) [reply packages/playground/windows/Playground-win32 (Package)/Playground-win32 (Package).wapproj:1]: @vmoroz  can you clarify what tools created for Unix might have to deal with a Windows app packaging project?

**Pull request merged**